src/core/border_router/routing_manager.cpp

```diff
@@ -1042,69 +1102,80 @@ bool RoutingManager::UpdateDiscoveredPrefixes(const RouterAdv::RouteInfoOption &
+    TimeMilli       now                      = TimerMilli::GetNow();
+    uint8_t         remainingOnLinkPrefixNum = 0;
+
+    mDiscoveredPrefixInvalidTimer.Stop();
 
     for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
         ExternalPrefix &prefix = mDiscoveredPrefixes[i];
 
         if ((aPrefix != nullptr && prefix.mPrefix == *aPrefix && prefix.mIsOnLinkPrefix == aIsOnLinkPrefix) ||
-            (prefix.mExpireTime <= now))
+            (prefix.GetExpireTime() <= now))
         {
             RemoveExternalRoute(prefix.mPrefix);
             ++removedNum;
         }
         else
         {
-            earliestExpireTime = OT_MIN(earliestExpireTime, prefix.mExpireTime);
-            *keptPrefix        = prefix;
-            ++keptPrefix;
+            mDiscoveredPrefixInvalidTimer.FireAtIfEarlier(prefix.GetExpireTime());
+            *remainingPrefix = prefix;
+            ++remainingPrefix;
             if (prefix.mIsOnLinkPrefix)
             {
-                ++keptOnLinkPrefixNum;
+                ++remainingOnLinkPrefixNum;
             }
         }
     }
 
     mDiscoveredPrefixNum -= removedNum;
 
-    if (keptOnLinkPrefixNum == 0)
-    {
-        mDiscoveredPrefixInvalidTimer.Stop();
-
-        if (mAdvertisedOnLinkPrefix == nullptr)
-        {
-            // There are no valid on-link prefixes on infra link now, start Router Solicitation
-            // To find out more on-link prefixes or timeout to advertise my local on-link prefix.
-            StartRouterSolicitationDelay();
-        }
-    }
-    else
+    if (remainingOnLinkPrefixNum == 0 && mAdvertisedOnLinkPrefix == nullptr)
     {
-        mDiscoveredPrefixInvalidTimer.FireAt(earliestExpireTime);
+        // There are no valid on-link prefixes on infra link now, start Router Solicitation
+        // To find out more on-link prefixes or timeout to advertise my local on-link prefix.
+        StartRouterSolicitationDelay();
     }
 
     return (removedNum != 0); // If anything was removed we need to reevaluate.
 }
 
 void RoutingManager::InvalidateAllDiscoveredPrefixes(void)
 {
-    TimeMilli past = TimerMilli::GetNow();
-
     for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
-        mDiscoveredPrefixes[i].mExpireTime = past;
+        mDiscoveredPrefixes[i].mValidLifetime = 0;
     }
 
     InvalidateDiscoveredPrefixes();
 
     OT_ASSERT(mDiscoveredPrefixNum == 0);
 }
 
+static void UpdateStaleTime(TimeMilli &aStaleTime,
+                            bool       aStaleTimeInitialized,
+                            bool       aIsOnLink,
+                            TimeMilli  aPrefixStaleTime)
+{
+    if (aIsOnLink)
+    {
+        if (!aStaleTimeInitialized || aPrefixStaleTime >= aStaleTime)
+        {
+            aStaleTime = aPrefixStaleTime;
+        }
+    }
+    else
+    {
+        if (!aStaleTimeInitialized || aPrefixStaleTime <= aStaleTime)
+        {
+            aStaleTime = aPrefixStaleTime;
```

Comment:
We add an external route for each OMR prefix in RIO and this consumes significant network data. This is for removing any OMR prefixes sooner to free network data space.

Reply:
The implementation has problem when there are both OMR prefixes and on-link prefixes, I updated to evaluate on-link prefix stale time and OMR prefix stale time separately.

At the same time, I think it may be beneficial to remove any on-link prefix as soon as we detected that the on-link prefix is no longer being advertised, for the same reason we remove the OMR prefixes. Thoughts?

Reply:
I think it depends on what we mean by "stale". Remember that just because we are no longer receiving RAs with the on-link prefix, does not mean that some host on the link still considers the on-link prefix as valid. We should be careful about removing an external route too quickly.